main/src/library/GetOpt.flix

```diff
@@ -0,0 +1,418 @@
+///
+/// Acknowledgement: this is a port of Sven Panne's `System.Console.GetOpt` in the Haskell Base libraries.
+///
+
+
+namespace GetOpt {
+
+    /// 
+    /// `ArgOrder` defines how to parse further arguments that follow non-options.
+    /// 
+    /// * `RequireOrder` after the first non-option all remain arguments are treated as non-options
+    /// * `Permute` options and non-options are freely interspersed in the input stream
+    /// * `ReturnInOrder` turn non-options into options by applying the supplied function
+    ///
+    pub enum ArgOrder[a] {
+        case RequireOrder
+        case Permute
+        case ReturnInOrder(String -> a)
+    }
+    
+    /// 
+    /// `ArgDescr` defines whether or not the option takes an argument.
+    /// 
+    /// * `NoArg` expects no argument. The constructor takes a value of type `a` indicating the option that has been recognized.
+    /// * `ReqArg` mandates an argument. The constructor takes a function to decode the argument and description string.
+    /// * `OptArg` optionally requires an argument. The constructor takes a function to decode the argument and description string.
+    ///
+    /// The "decode" functions are expected to produce `Some(_)` when decoding is successful, and `None` to indicate
+    /// failure.
+    ///
+    pub enum ArgDescr[a] {
+        case NoArg(a)
+        case ReqArg(String -> Option[a], String)
+        case OptArg(Option[String] -> Option[a], String)
+    }
+
+    ///
+    /// The kind of a command line option (internal).
+    ///
+    /// * `Opt(a)` is an option producing a value of type `a`
+    /// * `UnreqOp` is an unrecognized option
+    /// * `NonOpt` is a non-option, returned to the user as a String
+    /// * `EndOfOpts` signals the end of options in the decoded command line String
+    /// * `OptErr` is the error case with a String error message
+    ///    
+    enum OptionKind[a] {
+        case Opt(a)
+        case UnreqOpt(String)
+        case NonOpt(String)
+        case EndOfOpts
+        case OptErr(String)
+    }
+    
+    ///
+    /// An `OptionDescr` describes a single command line option. 
+    ///
+    /// * `optionIds` is a list of single character abbreviations identifying the option
+    /// * `optionNames` is a list of long names identifying the option
+    /// * `argDescriptor` defines the format of the option
+    /// * `explanation` is the description of the option printed to the user by the function `usageInfo`
+    ///
+    type alias OptionDescr[a] =
+        { optionIds: List[Char]
+        , optionNames: List[String]
+        , argDescriptor: ArgDescr[a]
+        , explanation: String
+        }
+
+
+    ///
+    /// Return a formatted string describing the usage of the command.
+    ///
+    /// Typically this output is used by "--help"
+    /// 
+    pub def usageInfo(header: String, optionDescriptors: List[OptionDescr[a]]): String = 
+        let paste = (x, y, z) -> "  ${x}  ${y}  ${z}";

```

Comment:
Are there two spaces here? is that intended?

Reply:
There are two spaces in the Haskell original, so I kept the same formatting.